File: imagepipeline/src/main/java/com/facebook/imagepipeline/memory/BufferMemoryChunk.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.imagepipeline.memory;

import android.util.Log;
import com.facebook.common.internal.Preconditions;
import java.io.Closeable;
import java.nio.ByteBuffer;
import javax.annotation.Nullable;

/**
 * Wrapper around chunk using a direct ByteBuffer in native memory. A direct ByteBuffer is composed
 * of a Java {@link ByteBuffer} object allocated on the Java heap and the underlying buffer which is
 * in native memory.
 *
 * <p>The buffer in native memory will be released when the Java object gets garbage collected.
 */
public class BufferMemoryChunk implements MemoryChunk, Closeable {
  private static final String TAG = "BufferMemoryChunk";

  /** Internal representation of the chunk */
  private ByteBuffer mBuffer;

  /** Size of the ByteBuffer */
  private final int mSize;

  /** Unique identifier of the chunk */
  private final long mId;

  public BufferMemoryChunk(final int size) {
    mBuffer = ByteBuffer.allocateDirect(size);
    mSize = size;
    mId = System.identityHashCode(this);
  }

  @Override
  public synchronized void close() {
    mBuffer = null;
  }

  @Override
  public synchronized boolean isClosed() {
    return mBuffer == null;
  }

  @Override
  public int getSize() {
    return mSize;
  }

  @Override
  public synchronized int write(
      final int memoryOffset, final byte[] byteArray, final int byteArrayOffset, final int count) {
    Preconditions.checkNotNull(byteArray);
    Preconditions.checkState(!isClosed());
    final int actualCount = MemoryChunkUtil.adjustByteCount(memoryOffset, count, mSize);
    MemoryChunkUtil.checkBounds(
        memoryOffset, byteArray.length, byteArrayOffset, actualCount, mSize);
    mBuffer.position(memoryOffset);
    mBuffer.put(byteArray, byteArrayOffset, actualCount);
    return actualCount;
  }

  @Override
  public synchronized int read(
      final int memoryOffset, final byte[] byteArray, final int byteArrayOffset, final int count) {
    Preconditions.checkNotNull(byteArray);
    Preconditions.checkState(!isClosed());
    final int actualCount = MemoryChunkUtil.adjustByteCount(memoryOffset, count, mSize);
    MemoryChunkUtil.checkBounds(
        memoryOffset, byteArray.length, byteArrayOffset, actualCount, mSize);
    mBuffer.position(memoryOffset);
    mBuffer.get(byteArray, byteArrayOffset, actualCount);
    return actualCount;
  }

  @Override
  public synchronized byte read(final int offset) {
    Preconditions.checkState(!isClosed());
    Preconditions.checkArgument(offset >= 0);
    Preconditions.checkArgument(offset < mSize);
    return mBuffer.get(offset);
  }

  @Override
  public void copy(
      final int offset, final MemoryChunk other, final int otherOffset, final int count) {
    Preconditions.checkNotNull(other);
    // This implementation acquires locks on this and other objects and then delegates to
    // doCopy which does actual copy. In order to avoid deadlocks we have to establish some linear
    // order on all BufferMemoryChunks and acquire locks according to this order. In order
    // to do that, we use unique ids.
    // So we have to address 3 cases:

    // Case 1: other buffer equals this buffer, id comparison
    if (other.getUniqueId() == getUniqueId()) {
      // we do not allow copying to the same address
      // lets log warning and not copy
      Log.w(
          TAG,
          "Copying from BufferMemoryChunk "
              + Long.toHexString(getUniqueId())
              + " to BufferMemoryChunk "
              + Long.toHexString(other.getUniqueId())
              + " which are the same ");
      Preconditions.checkArgument(false);
    }

    // Case 2: Other memory chunk id < this memory chunk id
    if (other.getUniqueId() < getUniqueId()) {
      synchronized (other) {
        synchronized (this) {
          doCopy(offset, other, otherOffset, count);
        }
      }
      return;
    }

    // Case 3: Other memory chunk id > this memory chunk id
    synchronized (this) {
      synchronized (other) {
        doCopy(offset, other, otherOffset, count);
      }
    }
  }

  @Override
  public long getNativePtr() {
    throw new UnsupportedOperationException("Cannot get the pointer of a BufferMemoryChunk");
  }

  @Override
  @Nullable
  public synchronized ByteBuffer getByteBuffer() {
    return mBuffer;
  }

  @Override
  public long getUniqueId() {
    return mId;
  }

  /**
   * This does actual copy. It should be called only when we hold locks on both this and other
   * objects
   */
  private void doCopy(
      final int offset, final MemoryChunk other, final int otherOffset, final int count) {
    if (!(other instanceof BufferMemoryChunk)) {
      throw new IllegalArgumentException("Cannot copy two incompatible MemoryChunks");
    }
    Preconditions.checkState(!isClosed());
    Preconditions.checkState(!other.isClosed());
    MemoryChunkUtil.checkBounds(offset, other.getSize(), otherOffset, count, mSize);
    mBuffer.position(offset);
    // ByteBuffer can't be null at this point
    other.getByteBuffer().position(otherOffset);
    // Recover the necessary part to be copied as a byte array.
    // This requires a copy, for now there is not a more efficient alternative.
    byte[] b = new byte[count];
    mBuffer.get(b, 0, count);
    other.getByteBuffer().put(b, 0, count);
  }
}

File: memory-types/simple/src/main/java/com/facebook/imagepipeline/memory/BufferMemoryChunk.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.memory

import android.util.Log
import java.io.Closeable
import java.nio.ByteBuffer

/**
 * Wrapper around chunk using a direct ByteBuffer in native memory. A direct ByteBuffer is composed
 * of a Java [ByteBuffer] object allocated on the Java heap and the underlying buffer which is in
 * native memory.
 *
 * The buffer in native memory will be released when the Java object gets garbage collected.
 */
class BufferMemoryChunk(
    /** Size of the ByteBuffer */
    override val size: Int
) : MemoryChunk, Closeable {

  /** Internal representation of the chunk */
  @get:Synchronized
  override var byteBuffer: ByteBuffer?
    private set

  /** Unique identifier of the chunk */
  override val uniqueId: Long

  @Synchronized
  override fun close() {
    byteBuffer = null
  }

  @Synchronized override fun isClosed(): Boolean = byteBuffer == null

  @Synchronized
  override fun write(
      memoryOffset: Int,
      byteArray: ByteArray,
      byteArrayOffset: Int,
      count: Int,
  ): Int {
    checkNotNull(byteArray)
    check(!isClosed())
    checkNotNull(byteBuffer)
    val actualCount = MemoryChunkUtil.adjustByteCount(memoryOffset, count, size)
    MemoryChunkUtil.checkBounds(memoryOffset, byteArray.size, byteArrayOffset, actualCount, size)
    byteBuffer!!.position(memoryOffset)
    byteBuffer!!.put(byteArray, byteArrayOffset, actualCount)
    return actualCount
  }

  @Synchronized
  override fun read(
      memoryOffset: Int,
      byteArray: ByteArray,
      byteArrayOffset: Int,
      count: Int,
  ): Int {
    checkNotNull(byteArray)
    check(!isClosed())
    checkNotNull(byteBuffer)
    val actualCount = MemoryChunkUtil.adjustByteCount(memoryOffset, count, size)
    MemoryChunkUtil.checkBounds(memoryOffset, byteArray.size, byteArrayOffset, actualCount, size)
    byteBuffer!!.position(memoryOffset)
    byteBuffer!![byteArray, byteArrayOffset, actualCount]
    return actualCount
  }

  @Synchronized
  override fun read(offset: Int): Byte {
    check(!isClosed())
    require(offset >= 0)
    require(offset < size)
    checkNotNull(byteBuffer)
    return byteBuffer!![offset]
  }

  override fun copy(offset: Int, other: MemoryChunk, otherOffset: Int, count: Int) {
    checkNotNull(other)

    // This implementation acquires locks on this and other objects and then delegates to
    // doCopy which does actual copy. In order to avoid deadlocks we have to establish some linear
    // order on all BufferMemoryChunks and acquire locks according to this order. In order
    // to do that, we use unique ids.
    // So we have to address 3 cases:

    // Case 1: other buffer equals this buffer, id comparison
    if (other.uniqueId == uniqueId) {
      // we do not allow copying to the same address
      // lets log warning and not copy
      Log.w(
          TAG,
          ("Copying from BufferMemoryChunk ${java.lang.Long.toHexString(uniqueId)} to BufferMemoryChunk ${java.lang.Long.toHexString(other.uniqueId)} which are the same "),
      )
      require(false)
    }

    // Case 2: Other memory chunk id < this memory chunk id
    if (other.uniqueId < uniqueId) {
      synchronized(other) { synchronized(this) { doCopy(offset, other, otherOffset, count) } }
      return
    }

    // Case 3: Other memory chunk id > this memory chunk id
    synchronized(this) { synchronized(other) { doCopy(offset, other, otherOffset, count) } }
  }

  override val nativePtr: Long
    get() {
      throw UnsupportedOperationException("Cannot get the pointer of a BufferMemoryChunk")
    }

  init {
    byteBuffer = ByteBuffer.allocateDirect(size)
    this.uniqueId = System.identityHashCode(this).toLong()
  }

  /**
   * This does actual copy. It should be called only when we hold locks on both this and other
   * objects
   */
  private fun doCopy(offset: Int, other: MemoryChunk, otherOffset: Int, count: Int) {
    require(other is BufferMemoryChunk) { "Cannot copy two incompatible MemoryChunks" }
    check(!isClosed())
    check(!other.isClosed())
    checkNotNull(byteBuffer)
    MemoryChunkUtil.checkBounds(offset, other.size, otherOffset, count, size)
    byteBuffer!!.position(offset)
    val otherByteBuffer = checkNotNull(other.byteBuffer)
    otherByteBuffer.position(otherOffset)
    // Recover the necessary part to be copied as a byte array.
    // This requires a copy, for now there is not a more efficient alternative.
    val b = ByteArray(count)
    byteBuffer!![b, 0, count]
    otherByteBuffer.put(b, 0, count)
  }

  companion object {
    private const val TAG = "BufferMemoryChunk"
  }
}
